src/app/shared/components/chip-autocomplete/component.ts

```diff
@@ -0,0 +1,194 @@
+import {Subject} from 'rxjs';
+import {map, startWith, takeUntil} from 'rxjs/operators';
+
+enum Controls {
+  Tags = 'tags',
+  Filter = 'filter',
+}
+
+/**
+ * @title Chips Autocomplete
+ */
+@Component({
+  selector: 'km-chip-autocomplete',
+  templateUrl: './template.html',
+  styleUrls: ['./style.scss'],
+  providers: [
+    {
+      provide: NG_VALUE_ACCESSOR,
+      useExisting: forwardRef(() => ChipAutocompleteComponent),
+      multi: true,
+    },
+    {
+      provide: NG_VALIDATORS,
+      useExisting: forwardRef(() => ChipAutocompleteComponent),
+      multi: true,
+    },
+  ],
+})
+export class ChipAutocompleteComponent implements OnChanges, OnInit, OnDestroy, ControlValueAccessor, Validator {
+  readonly Controls = Controls;
+
+  form: FormGroup;
+  allTags: string[] = [];
+  filteredTags: string[] = [];
+  selectedTags: string[] = [];
+  separatorKeysCodes: number[] = [ENTER, COMMA, SPACE];
+
+  @Input() label: string;
+  @Input() title: string;
+  @Input() tags: string[] = [];
+  @Input() description = 'Use comma, enter or space key as the separator.';
+  @Input() placeholder = 'Select single or multiple values';
+  @Input() required = false;
+
+  // Note: Event to emit selected values in order to use this component as independent component.
+  @Output() Change = new EventEmitter<string[]>();
+  @ViewChild('tagInput') tagInput: any;
+
+  private _unsubscribe = new Subject<void>();
+
+  constructor(private readonly _builder: FormBuilder) {}
+
+  ngOnChanges(changes: SimpleChanges): void {
+    if (changes && changes.tags) {
+      this.allTags = changes.tags.currentValue;
+      this.filteredTags = this.allTags;
+
+      if (this.allTags.length === 0) {
+        this.selectedTags = [];
+      }
+    }
+  }
+
+  ngOnInit(): void {
+    this._initForm();
+    this._initSubscriptions();
+  }
+
+  ngOnDestroy(): void {
+    this._unsubscribe.next();
+    this._unsubscribe.complete();
+  }
+
+  addTag(event: MatChipInputEvent): void {
+    const value = (event.value || '').trim();
+    if (value) {
+      this.selectedTags.push(value);
+    }
+    // Clear the input value
+    event.chipInput!.clear();
+    this._patchValue();
+    this.Change.emit(this.selectedTags);
+  }
+
+  removeTag(tag: string): void {
+    const index = this.selectedTags.indexOf(tag);
+    if (index >= 0) {
+      this.selectedTags.splice(index, 1);
+    }
+    this._patchValue();
+    this.filteredTags = this._removeSelectedTagsFromList();
+    this.Change.emit(this.selectedTags);
+  }
+
+  selected(event: MatAutocompleteSelectedEvent): void {
+    this.selectedTags.push(event.option.viewValue);
+    this.tagInput.nativeElement.value = '';
+    this._patchValue();
+    this.filteredTags = this._removeSelectedTagsFromList();
+    this.Change.emit(this.selectedTags);
+  }
+
+  validate(_: AbstractControl): ValidationErrors {
+    return this.form.get(Controls.Tags)?.valid ? null : this.form.get(Controls.Tags)?.errors;
```

Comment:
when the control will be `valid`, `this.form.get(Controls.Tags)?.errors` will return `null` there is no need to check if the control is `valid` is not.